libr/core/canal.c

```diff
@@ -2769,12 +2779,17 @@ static bool anal_path_exists(RCore *core, ut64 from, ut64 to, RList *bbs, int de
 	if (!bb || !bbs) {
 		return false;
 	}
-	r_list_append (state, (void *)from);
+
+	*frp = from;
+	r_list_append (state, (void *)frp);
+
+	ut64 tj = bb->jump;
+	ut64 fj = bb->addr + bb->size;
 
 	// try to find the target in the current function
 	if (r_anal_bb_is_in_offset (bb, to) ||
-			((!r_list_contains (state, (void *)bb->jump) && anal_path_exists (core, bb->jump, to, bbs, depth - 1, state))) ||
-			((!r_list_contains (state, (void *)(bb->addr + bb->size)) && anal_path_exists (core, bb->addr + bb->size, to, bbs, depth - 1, state)))) {
+			((!r_list_find (state, (void *)&(tj), cmp_addr) && anal_path_exists (core, bb->jump, to, bbs, depth - 1, state))) ||
```

Comment:
may be better to use r_ht_up, there seems to be many places where this list is iterated, now that hashtables are smaller we can evaluate using them in here. also, im a bit afraid of this hacky void*/ut64 usage. which would is solved if rewritten with the hashtable

Reply:
yes the hash table is the best thing for this use case. See `sdb/ht_uu.h` and `sdb/ht_uu.c`

Reply:
ok ! will do that.

Reply:
just to make sure you didn't miss that change, I stopped casting ut64 to void* by using `r_list_find` instead of r_list_contains. and allocating a new ut64 each time I need to...
https://github.com/radare/radare2/blob/db2a8619761c76dcc9f6aa3bef9c5dde4e2e7315/libr/core/canal.c#L2772
https://github.com/radare/radare2/blob/db2a8619761c76dcc9f6aa3bef9c5dde4e2e7315/libr/core/canal.c#L2783
https://github.com/radare/radare2/blob/db2a8619761c76dcc9f6aa3bef9c5dde4e2e7315/libr/core/canal.c#L2756

but no problem can switch to hash tables, but isn't there an equivalent of sets instead ? because I just need unique values. not k/v pairs.

Reply:
No we don't have `set` yet, but you can easily use `HtUU` easily to simulate a set. You just need to put your elements as keys (no need to allocate them on the heap) and a dummy value, just to mark them as present.

Reply:
done